internal/gitinterface/repository.go

```diff
@@ -104,7 +104,47 @@ func (r *Repository) executeGitCommandDirectString(args ...string) (string, erro
 // repository. It executes in the current directory without specifying the
 // GIT_DIR explicitly.
 func (r *Repository) executeGitCommandDirect(args ...string) (io.Reader, io.Reader, error) {
+	return r.executeGitCommandDirectWithEnv(nil, args...)
+}
+
+// executeGitCommandWithEnvString is a helper to execute the specified command
+// in the repository after setting the provided environment variables. Note that
+// the command inherits os.Environ() first. This helper explicitly sets the
+// GIT_DIR. It returns the stdout for successful execution as a string, with
+// leading and trailing spaces removed.
+func (r *Repository) executeGitCommandWithEnvString(env []string, args ...string) (string, error) {
+	stdOut, stdErr, err := r.executeGitCommandWithEnv(env, args...)
+	if err != nil {
+		stdErrContents, newErr := io.ReadAll(stdErr)
+		if newErr != nil {
+			return "", fmt.Errorf("unable to read stderr contents: %w; original err: %w", newErr, err)
+		}
+		return "", fmt.Errorf("%w when executing `git %s`: %s", err, strings.Join(args, " "), string(stdErrContents))
+	}
+
+	stdOutContents, err := io.ReadAll(stdOut)
+	if err != nil {
+		return "", fmt.Errorf("unable to read stdout contents: %w", err)
+	}
+
+	return strings.TrimSpace(string(stdOutContents)), nil
+}
+
+// executeGitCommandWithEnv is a helper to execute the specified command in the
+// repository after setting the provided environment variables. Note that the
+// command inherits os.Environ() first. This helper explicitly sets the GIT_DIR.
+func (r *Repository) executeGitCommandWithEnv(env []string, args ...string) (io.Reader, io.Reader, error) {
+	args = append([]string{"--git-dir", r.gitDirPath}, args...)
+	return r.executeGitCommandDirectWithEnv(env, args...)
+}
+
+// executeGitCommandDirectWithEnv is a helper to execute the specified command
+// in the repository after setting the provided environment variables. Note that
+// the command inherits os.Environ() first. This helper executes in the current
+// directory without specifying the GIT_DIR explicitly.
+func (r *Repository) executeGitCommandDirectWithEnv(env []string, args ...string) (io.Reader, io.Reader, error) {
 	cmd := exec.Command(binary, args...)
+	cmd.Env = append(os.Environ(), env...)
```

Comment:
Up to you if you want to do this in this PR or not, but having different funcs for different modes of operation is going to be painful to maintain in the long run (e.g. look how many places we set `--git-dir`). It might be better if we follow the approach of exec.Command and break this into it's own type that can be modified.

e.g.

```go
cmd := r.NewCommand("commit", ...)
cmd.Env = map[string]string{...}
stdout, stderr, err := cmd.Do()
```

or if we want to get fancier (depending on how much validation we need to do)

```go
stdout, stderr, err := r.NewCommand("commit", ...).WithEnv(...).Do()
```

Reply:
Agreed, this was already getting to be a pain to manage / track all the different combinations. I'll update this separately, good suggestion!